tests/ui/transmute/transmute-zst-generics.rs

```diff
@@ -0,0 +1,14 @@
+// run-pass
+// Transmuting to/from ZSTs that contain generics.
+
+#![feature(transmute_generic_consts)]
+
+unsafe fn cast_zst<T>(from: ()) -> [T; 0] {
+    ::std::mem::transmute::<(), [T; 0]>(from)
+}
```

Comment:
does this work the other direction? not saying that it needs to, just curious

Reply:
Yes, it does! I extended the test-case in my local branch to show that it works both directions, works with compound generic types, and that it propagates through arrays (but not structs, etc.):

```rust
// run-pass
// Transmuting to/from ZSTs that contain generics.

#![feature(transmute_generic_consts)]

// Verify non-generic ZST -> generic ZST transmute
unsafe fn cast_zst0<T>(from: ()) -> [T; 0] {
    ::std::mem::transmute::<(), [T; 0]>(from)
}

// Verify generic ZST -> non-generic ZST transmute
unsafe fn cast_zst1<T>(from: [T; 0]) -> () {
    ::std::mem::transmute::<[T; 0], ()>(from)
}

// Verify transmute with generic compound types
unsafe fn cast_zst2<T>(from: ()) -> [(T, T); 0] {
    ::std::mem::transmute::<(), [(T, T); 0]>(from)
}

// Verify transmute with ZST propagation through arrays
unsafe fn cast_zst3<T>(from: ()) -> [[T; 0]; 8] {
    ::std::mem::transmute::<(), [[T; 0]; 8]>(from)
}

pub fn main() {
    unsafe {
        let _: [u32; 0] = cast_zst0(());
        let _ = cast_zst1::<u32>([]);
        let _: [(u32, u32); 0] = cast_zst2(());
        let _: [[u32; 0]; 8] = cast_zst3(());
    };
}
```

Reply:
I think it's probably good to have these tests pushed as well. It may also be worth showing that `[T; N]` can be transmuted, and at the callsite `N=0`.

Reply:
Yeah, I will push the extended tests later. I am playing with struct propagation as well. But I don't think you can transmute `[T; N]`. The transmute check is done on the generic code, and it cannot show that it is valid for all `N`.

Reply:
Ah true, it depends if the target struct is also generic over `N`, but not sure if that's a trivial case or not.